File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = head;
        int len = 0;
        while (temp != nullptr) {
            len++;
            temp= temp->next;
        }

        int k = len - n + 1;
        temp = head;
        int curr = 1;
        while (curr < (k - 1)) {
            temp = temp->next;
            curr++;
        }

        if (temp->next == nullptr) {
            return nullptr;
        }

        if (k == 1) {
            head = temp->next;
            return head;
        }


        temp->next = temp->next->next;
        return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) return head;

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode curr = head;
        ListNode tail = dummy;
        
        for (int i = 1; i <= n; i++) {
            curr = curr.next;
        }

        while (curr != null) {
            curr = curr.next;
            tail = tail.next;
        }

        tail.next = tail.next.next;
        return dummy.next;
    }
}